### ISSUE TITLE
Template format error: Unresolved resource dependencies [*bucket-name*] in the Resources block of the template

### ISSUE THREAD

**Issue by pierskarsenbarg** [serverless/serverless#2976]
# This is a Bug Report

## Description

For bug reports:
* What went wrong? I got the error message  **Template format error: Unresolved resource dependencies [pk-bucket] in the Resources block of the template**
* What did you expect should have happened? An IAM policy should be created that allows access from lambda to S3 bucket
* What was the config you used?

```yaml
resources:
  Resources:
    S3Bucket:
      Type: AWS::S3::Bucket
      Properties:
        BucketName: pk-bucket
    myDefaultRole:
      Type: AWS::IAM::Role
      Properties:
        Path: /my/default/path
        RoleName: MyDefaultRole
        AssumeRolePolicyDocument:
          Version: '2012-10-17'
          Statement:
            - Effect: Allow
              Principal:
                Service:
                  - lambda.amazonaws.com
              Action: sts:AssumeRole
        Policies:
          - PolicyName: myPolicyName
            PolicyDocument:
              Version: '2012-10-17'
              Statement:                
                -  Effect: "Allow"
                   Action:
                     - "s3:PutObject"
                   Resource:
                     Fn::Join:
                       - ""
                       - - "arn:aws:s3:::"
                         - "Ref" : "pk-bucket"
```

* What stacktrace or error message from your provider did you see? See above

## Additional Data

* ***Serverless Framework Version you're using***: 1.4
* ***Operating System***: Windows 10
* ***Stack Trace***:
* ***Provider Error messages***:

**Comment by eahefnawy** [comment]
@pierskarsenbarg I think you're not referencing the bucket correctly in the role template. I see that you're referencing by the bucket name:

```yml
- "Ref" : "pk-bucket"
```

You should instead reference by the logical ID of the bucket resource:

```yml
- "Ref" : "S3Bucket"
```

**Comment by pierskarsenbarg** [comment]
Yup that worked. Thank you.

**Comment by rahul9580** [comment]
could you please let me know what is the logical ID of the bucket resource?

**Comment by pmuens** [comment]
Hey @rahul9580 the schema which is followed for the resource logical id generation can be found in this table: https://serverless.com/framework/docs/providers/aws/guide/resources#aws-cloudformation-resource-reference

**Comment by sola93** [comment]
Hi, I'm trying to add iamRoleStatements, and in the Ref I'm writing the following:
`iamRoleStatements:
    - Effect: "Allow"
      Action:
        - "s3:GetObject"
      Resource: { "Fn::Join" : ["", ["arn:aws:s3:::", { "Ref" : "S3BucketBUCKER_NAME" } ] ]  }`

But I'm getting the same error. What Am I doing wrong?

**Comment by jchan172** [comment]
I ran into this issue when I saved an Elastic Beanstalk configuration and then tried to launch an environment with the saved configuration without modifications, and I got the error `Service:AmazonCloudFormation, Message:Template format error: Unresolved resource dependencies [AWSEBLoadBalancerSecurityGroup] in the Resources block of the template`. Trying again, I clicked Launch Environment, but this time I clicked to see more configuration options, and I saw the problem. The load balancer was not selected properly (I had the Classic load balancer in my configuration, but it chose the default Application Load Balancer). So, it seems the Elastic Beanstalk Console/UI doesn't select the load balancer from the configuration file. I had to click to switch to the Classic load balancer manually before things worked fine.
